lib/ui/painting/scene/scene_node.cc

```diff
@@ -0,0 +1,142 @@
+#include "third_party/tonic/dart_library_natives.h"
+
+namespace flutter {
+
+IMPLEMENT_WRAPPERTYPEINFO(ui, SceneNode);
+
+void SceneNode::Create(Dart_Handle wrapper) {
+  auto res = fml::MakeRefCounted<SceneNode>();
+  res->AssociateWithDartWrapper(wrapper);
+}
+
+std::string SceneNode::initFromAsset(const std::string& asset_name,
+                                     Dart_Handle completion_callback_handle) {
+  FML_TRACE_EVENT("flutter", "SceneNode::initFromAsset", "asset", asset_name);
+
+  if (!Dart_IsClosure(completion_callback_handle)) {
+    return "Completion callback must be a function.";
+  }
+
+  auto dart_state = UIDartState::Current();
+  if (!dart_state->IsImpellerEnabled()) {
+    return "3D scenes require the Impeller rendering backend to be enabled.";
+  }
+
+  std::shared_ptr<AssetManager> asset_manager =
+      dart_state->platform_configuration()->client()->GetAssetManager();
+  std::unique_ptr<fml::Mapping> data = asset_manager->GetAsMapping(asset_name);
+  if (data == nullptr) {
+    return std::string("Asset '") + asset_name + std::string("' not found.");
+  }
+
+  auto& task_runners = dart_state->GetTaskRunners();
+
+  std::promise<std::shared_ptr<impeller::Context>> context_promise;
+  auto impeller_context_promise = context_promise.get_future();
+  task_runners.GetIOTaskRunner()->PostTask(
+      fml::MakeCopyable([promise = std::move(context_promise),
+                         io_manager = dart_state->GetIOManager()]() mutable {
+        promise.set_value(io_manager ? io_manager->GetImpellerContext()
+                                     : nullptr);
+      }));
+
+  auto persistent_completion_callback =
+      std::make_unique<tonic::DartPersistentValue>(dart_state,
+                                                   completion_callback_handle);
+
+  auto ui_task = fml::MakeCopyable(
+      [this, callback = std::move(persistent_completion_callback)](
+          std::shared_ptr<impeller::scene::Node> node) mutable {
+        auto dart_state = callback->dart_state().lock();
+        if (!dart_state) {
+          // The root isolate could have died in the meantime.
+          return;
+        }
+        tonic::DartState::Scope scope(dart_state);
+
+        node_ = std::move(node);
+        tonic::DartInvoke(callback->Get(), {Dart_TypeVoid()});
+
+        // callback is associated with the Dart isolate and must be
+        // deleted on the UI thread.
+        callback.reset();
+      });
+
+  task_runners.GetRasterTaskRunner()->PostTask(
+      fml::MakeCopyable([ui_task = std::move(ui_task), task_runners,
+                         impeller_context = impeller_context_promise.get(),
+                         data = std::move(data)]() {
+        auto node = impeller::scene::Node::MakeFromFlatbuffer(
+            *data, *impeller_context->GetResourceAllocator());
+
+        task_runners.GetUITaskRunner()->PostTask(
+            [ui_task, node = std::move(node)]() { ui_task(node); });
+      }));
```

Comment:
@chinmaygarde Could you spot check the threading code above? 

I pulled the dart state persistence stuff from the image decoding routine when I ran into trouble with it. I think I know how it works now, but I'm not super experienced with the Dart runtime's behavior.

Also, I'm fetching the Impeller context from an IO ref because that's all I happen to have found. Is there an easily accessible ref I can grab on the raster thread instead?

Reply:
So, I am not sure why the scene needs to be created on the raster task runner. That is just going to cause jank. We should decompress it on the concurrent worker pool like the texture decompression. But this is fine for now and safe.

Reply:
Huh, I guess I haven't verified the safety of doing it off the raster thread. In the moment I thought the geometry + texture buffer uploads would fall over for GLES because we don't manage multiple contexts... but actually I'm pretty sure that's not an issue because IIRC we enqueue everything in the reactor and then execute it on the raster thread later anyways.

Reply:
Yeah. I want to move away from the model of certain objects being only safely accessible from specific threads. If there is missing synchronization, that is a definitely a construction issue and not one that should restrict the use of the object from any thread.

You're right about the OpenGL ES stuff using the reactor.